.NET/Microsoft.Recognizers.Text.DateTime/Parsers/CJK/BaseCJKDatePeriodParser.cs

```diff
@@ -0,0 +1,1622 @@
+                return HandleYearResult(ret, hasHalf, isFirstHalf, year);
+            }
+
+            match = this.config.YearInChineseRegex.MatchExact(text, trim: true);
+
+            if (match.Success)
+            {
+                var yearStr = match.Value;
+
+                // Handle like "二零一七年上半年"，"二零一七年下半年"
+                yearStr = HandleWithHalfYear(match, yearStr, out bool hasHalf, out bool isFirstHalf);
+
+                if (this.config.IsYearOnly(yearStr))
+                {
+                    yearStr = yearStr.Substring(0, yearStr.Length - 1);
+                }
+
+                if (yearStr.Length == 1)
+                {
+                    return ret;
+                }
+
+                var year = ConvertChineseToInteger(yearStr);
+
+                return HandleYearResult(ret, hasHalf, isFirstHalf, year);
+            }
+
+            return ret;
+        }
+
+        private string HandleWithHalfYear(ConditionalMatch match, string text, out bool hasHalf, out bool isFirstHalf)
+        {
+            var firstHalf = match.Groups["firstHalf"].Value;
+            var secondHalf = match.Groups["secondHalf"].Value;
+
+            hasHalf = false;
+            isFirstHalf = !string.IsNullOrEmpty(firstHalf);
+
+            if (isFirstHalf || !string.IsNullOrEmpty(secondHalf))
+            {
+                var halfText = isFirstHalf ? firstHalf : secondHalf;
+                text = text.Substring(0, text.Length - halfText.Length);
+                hasHalf = true;
+            }
+
+            return text;
+        }
+
+        private DateTimeResolutionResult HandleYearResult(DateTimeResolutionResult ret, bool hasHalf, bool isFirstHalf, int year)
+        {
+            if (year < 100 && year >= this.config.TwoNumYear)
+            {
+                year += 1900;
+            }
+            else if (year < 100 && year < this.config.TwoNumYear)
+            {
+                year += 2000;
+            }
+
+            var beginDay = DateObject.MinValue.SafeCreateFromValue(year, 1, 1);
+            var endDay = DateObject.MinValue.SafeCreateFromValue(year + 1, 1, 1);
+
+            ret.Timex = DateTimeFormatUtil.LuisDate(year);
+
+            if (hasHalf)
+            {
+                if (isFirstHalf)
+                {
+                    endDay = DateObject.MinValue.SafeCreateFromValue(year, 7, 1);
+                }
+                else
+                {
+                    beginDay = DateObject.MinValue.SafeCreateFromValue(year, 7, 1);
+                }
+
+                ret.Timex = TimexUtility.GenerateDatePeriodTimex(beginDay, endDay, DatePeriodTimexType.ByMonth);
+            }
+
+            ret.FutureValue = ret.PastValue = new Tuple<DateObject, DateObject>(beginDay, endDay);
+            ret.Success = true;
+
+            return ret;
+        }
+
+        // parse entities that made up by two time points
+        private DateTimeResolutionResult MergeTwoTimePoints(string text, DateObject referenceDate)
```

Comment:
Is there something here that's CJK specific?

Reply:
it is a bit different from the standard version, for example it does not implement the following steps:
- ParseNowAsDate
- // Propagate the possible future relative context from the first entity to the second one in the range.
- // Check if weekPrefix is already included in the extractions otherwise include it
- // Expressions like "today", "tomorrow",... should keep their original year

I am not sure if these steps are required in the CJK implementation, and since the corresponding regexes are not defined in the Chinese yaml file, I have kept the two versions separate for now. Possibly we can merge them (and move them to some utils file) if it turns out to be necessary as we develop Japanese and Korean.

Reply:
Can you add a `@TODO` code comment here as a reminder of that?

Reply:
done